### PR TITLE
[BEAM-1113] Re-enable UsesTimersInParDo tests in Dataflow runner

### PR DESCRIPTION
**Pull request by kennknowles**
A rather vital commit left out of #1792. Do not review yet; utilizing Jenkins.

### CODE REVIEW

**Comment by kennknowles** [comment]
Run Dataflow RunableOnService

**Comment by kennknowles** [comment]
Rebased to pick up the updated worker.

**Comment by kennknowles** [comment]
Run Dataflow RunableOnService

**Comment by kennknowles** [comment]
The issue here was that we do not have a consistent spec versus on when a runner unwraps a `UserCodeException`, when a runner converts it to `PipelineExecutionException`. Our existing tests get around that by just requiring any sort of `RuntimeException` so I've made these tests match our existing tests.

**Comment by kennknowles** [comment]
Run Dataflow RunableOnService

**Comment by dhalperi** [comment]
@kennknowles -- your root cause makes sense, and confirmed via Jenkins output.

Sanity check -- @tgroh, since I think you thought hard about this last, can you confirm whether we've decided to try to make the way errors are surfaced standardized across runners?

And as another pointer, I recently dealt with this issue and decided that `ExpectedException` wasn't powerful enough. Instead, I just did something like this:

```java
try {
   ... // failing code, e.g., p.run().waitUntilFinish();
   fail(); // should not get here.
} catch (Exception e) {
  String fullTrace = Throwables.getStackTraceAsString(e);
  assertThat(fullTrace, ... /* stuff I wish ExpectedException did */);
}
```

**Comment by kennknowles** [comment]
Last time I worked on this, we actually decided to be lax about it because we don't have a short-term story for getting all runners to agree. I don't know that there's any design or consensus. Bytecode generation also changed exactly what was user code or not.

My favored style is a `PipelineExecutionException` with a cause that indicates whether the user actually threw the exception or the runner threw the exception, with a cause that is details. But that might not actually be friendly when dumped. Fn API will make this distinction quite stark so that seems the most expedient way towards a design. In a sense, the thing that has already bitten us many times is the distinction between "user exception" and "SDK harness exception" - the runner exceptions are already fairly well separated I think.

For this case, I'm not sure where the assertion failure is coming from (which is why it isn't a `RuntimeException`); I think the stack trace truncation might be hiding the details. I'd love to just follow the pattern we have already so we can save design work for later.

Changing the test to some other common superclass of exceptions would definitely work, but this feels solvable by more investigation.

**Comment by tgroh** [comment]
There's [a comment](https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java#L91) in `TestPipeline` that we must do so for `PAssert` failures. I don't believe there's been any work put into having consistent exceptions thrown from a non-assertion-based failure; part of that is due to differing retry behavior, flaky dependencies, etc (e.g. what if work fails in four different ways in Dataflow? which one "counts" as the one that failed the Pipeline?)

We also don't really have a strong contract on the behavior of `getState()` or `waitToFinish()`, at least at the `Pipeline` level - they only say ["Returning the Final State"](https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/PipelineResult.java#L62).

**Comment by kennknowles** [comment]
Run Dataflow RunnableOnService

**Comment by kennknowles** [comment]
Run Dataflow RunnableOnService

**Comment by kennknowles** [comment]
Run Dataflow RunnableOnService

**Comment by kennknowles** [comment]
Run Dataflow RunnableOnService

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java [hunk 1]
```diff
@@ -162,7 +162,7 @@ DataflowPipelineJob run(Pipeline pipeline, DataflowRunner runner) {
         throw new IllegalStateException(
             "The dataflow did not output a success or failure metric.");
       } else if (!success.get()) {
-        throw new AssertionError(
+        throw new RuntimeException(
```
Comment (kennknowles) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java]: This is actually not the right fix.

**Comment by kennknowles** [comment]
Run Dataflow RunnableOnService

**Comment by aaltay** [comment]
Hi @kennknowles, any updates on this PR?

**Comment by kennknowles** [comment]
No, I haven't had time to address the problems in TestDataflowRunner.

**Comment by davorbonaci** [comment]
@kennknowles, any updates perhaps?

**Comment by kennknowles** [comment]
Run Dataflow ValidatesRunner

**Comment by kennknowles** [comment]
There were other changes to `TestDataflowRunner`. I don't actually think they'll solve the problem; I believe it throws `AssertionError` spuriously still. I'll let Jenkins confirm/deny.

**Comment by kennknowles** [comment]
Run Dataflow ValidatesRunner

**Comment by kennknowles** [comment]
run dataflow validatesrunner

**Comment by kennknowles** [comment]
run Dataflow ValidatesRunner

**Comment by kennknowles** [comment]
run Dataflow ValidatesRunner

**Comment by kennknowles** [comment]
I can repro this offline, so I will reopen when I am looking for review.

**Pull request closed without merging**